### ISSUE TITLE
Idiomatic schematics for NestJS

### ISSUE THREAD

**Issue by beeman** [nrwl/nx#2959]
## Expected Behavior

It would be awesome if we could enhance the `@nrwl/nest` schematics to be more idiomatic to an Nx workspace, and support this syntax:

`nx generate @nrwl/nest:<schematic> --project <project>`. 

## Current Behavior

Currently we add `@nestjs/schematics` as a peer dependency and while that allows you to generate NestJS code, there are a few things with it.

Firstly, it doesn't have the notion of `projects`, so in order to generate code in one of our libraries we need to pass in the full path, like so:

```
$ nx g @nestjs/schematics:service todo --sourceRoot libs/todo/src/lib 
CREATE libs/todo/src/lib/todo/todo.service.spec.ts (446 bytes)
CREATE libs/todo/src/lib/todo/todo.service.ts (88 bytes)
UPDATE libs/todo/src/lib/todo.module.ts (194 bytes)
```

I would love to be able to just to `nx g @nrwl/nest:service --project todo` - similar to for instance `@nrwl/angular` or `@nrwl/react`.

Additionally, `@nestjs/schematics` has a couple of schematics that don't really make sense in the context of a Nx Workspace:

- angular-app
- application
- configuration
- library
- sub-app

### Context

To start, I think it makes sense to port over some of the schematics that are most frequently used (at least by me that is, your milage may vary). 

- [ ] controller
- [ ] module
- [ ] resolver
- [ ] service

These are some other schematics from `@nestjs/schematics` that might be good to port at some point:

- [ ] class
- [ ] decorator
- [ ] filter
- [ ] gateway
- [ ] guard
- [ ] interceptor
- [ ] interface
- [ ] middleware
- [ ] pipe
- [ ] provider

Lastly, I'd love to add a schematic that allows configuring GraphQL in a nest module. So `nx g @nrwl/nest:graphql --project api` would install and configure `@nestjs/graphql` in the `api` project.

## Other

I'm opening this issue to see if it makes sense to add this. If so, I'd be happy to create a PR that adds the first 4 schematics I listed above, and later optionally ad some more. 

Also, as I see myself using these schematics quite a bit in the future, I'm happy to take on maintaining these schematics inside `nrwl/nx` for the foreseeable future.

**Comment by Cammisuli** [comment]
Hi @beeman, we need to think of a way to dynamically map all the nest schematics to support the `--project` flag. Currently we would have to do what you suggested and have a one to one mapping between nest and nrwl. And there's a lot. It'll be a big maintenance burden ☹️ 

This is something that we need to talk about more for sure. 

In the meantime, you can create workspace schematics that map the ones you need for your repo.

As for the graphql one, you're more than welcomed to add support for that 😄 . If you need help with it, you can reach out and I'll be more than happy to help you.

**Comment by beeman** [comment]
@Cammisuli thanks for your feedback. It makes total sense to check if we can dynamically invoke `@nestjs/schematics` and inject the correct path. I will take a look to see if it's low-hanging fruit.

With regards to the GraphQL schematic, I started working on an implementation in #2961, I would love to get some early feedback :)

**Comment by beeman** [comment]
@Cammisuli I created another PR that wraps around `@nestjs/schematics:module`, not sure if this is what you meant with mapping around the existing schematics?

Or is there a way we can be more dynamic?

**Comment by Cammisuli** [comment]
I was thinking more dynamic. Where we don't have to provide our own schematic for each of nests. It's definitely a ponderer 🤔

**Comment by beeman** [comment]
I would love to do some more experiments to see if that's possible, do you have an api in mind for the end user? What a command would look like?

The main issue I see with having it fully dynamic is that the values in schema.json are hardcoded but different between various schematics.

**Comment by FrozenPandaz** [comment]
Can we write this function?

```
// T is some options from `@nestjs/schematics`
nestSchematic<T = any>(name: string, options: Exclude<T, 'sourceRoot'> & { project: string }) {
  const nestOptions: T = convertToNestOptions(options);
  externalSchematics('@nestjs/schematics', name, nestOptions);
}
```

**Comment by yharaskrik** [comment]
Not sure what the status of this is but @FrozenPandaz I suggested the same thing to @Cammisuli a while ago. Convert the nx expected flags to nest CLI expected flags so that people familiar with nx can use the nest CLI without learning something else. I'm open to working on this and creating thar convert function if you want? Can add it into part of the nestjs lib functionality I wrote that was merged into v9.
